### DISCUSSION TITLE
Livewire encountered corrupt data when trying to hydrate on first load

### DISCUSSION THREAD

**Discussion by D-Magic** [livewire/livewire#4753] [1. Help]
Hello,

Normally our components are loading fine, but sometimes random component fails because of "Livewire encountered corrupt data when trying to hydrate" error.

This is only happening on the initial load of the component so it doesn't display / we get the error 500 popup.

We are struggling to understand why the DOM is changing as the component hasn't loaded yet.

Struggling to debug this as we also see some elements end up with a wire:initial-data attribute and some don't.

What would be ideal is if there is a way to trap this in the rare event and reload the component?

Only reason we are focusing on this is this particular site has tens of thousands of users per day so we are getting several hundred of these errors per day.

It just seems like a simple refresh will resolve but unsure how to handle this?

Please see an example of one of the components that fails sometimes:

```php
<?php

namespace Denfield\Reviews\Http\Livewire\Reviews;

use Denfield\Reviews\Repositories\Contracts\ReviewsRepositoryContract;
use Illuminate\Contracts\View\View;
use Livewire\Component;

class Review extends Component
{
    /**
     * @var string|null
     */
    public ?string $class = null;

    /**
     * @var bool
     */
    public bool $hideIcons = false;

    /**
     * @var bool
     */
    public bool $hideRating = false;

    /**
     * @var bool
     */
    public bool $hideReviews = false;

    /**
     * @var string|null
     */
    public ?string $iconSize = null;

    /**
     * @var bool
     */
    public bool $loading = true;

    /**
     * @var array
     */
    public array $meta = [];

    /**
     * @var string|null
     */
    public ?string $resource = null;

    /**
     * @var string|null
     */
    public ?string $resourceId = null;

    /**
     * @var array|null
     */
    public ?array $review = null;

    public function init()
    {
        try {
            $this->review = !$this->resource
                ? app( ReviewsRepositoryContract::class )->total( $this->resourceId )()
                : app( ReviewsRepositoryContract::class )->resource( $this->resource, $this->resourceId )();
        } catch ( \Exception $e ) {
        }

        $this->loading = false;
    }

    /**
     * @return View
     */
    public function render(): View
    {
        return view( "livewire.reviews." . ( !$this->resource ? 'review' : 'widget' ) );
    }
}
```

Kind regards,
Maciej

**Comment by joshhanley** [comment]
@D-Magic something else must be happening as there is no way to get the corrupt data error on page load as it only does the check when a component is re-hydrated as part of a subsequent network request.

If you say you are getting a 500 error on page load, how do you know that you are getting the corrupt data error? Is it possible that you are getting 500 errors on some page loads but the corrupt data is happening in different components after a subsequent network request has happened?

`wire:initial-data` only remains on the element until the component has been loaded in javascript, then it is removed from the element. If you are still seeing that, it suggests that something is erroring on the javascript side when a component tries to load. Do you have an errors in your browser console?

**Reply by D-Magic** [reply]
Thank you for your response. There is 500 error in live environment but I can get error from Sentry.io.

Endpoint: `/livewire/message/automotive-vehicles-listing`

```
Symfony\Component\HttpKernel\Exception\HttpException: Livewire encountered corrupt data when trying to hydrate the [denfield.automotive.http.livewire.vehicles.listing] component. 
Ensure that the [name, id, data] of the Livewire component wasn't tampered with between requests.
```

This error is always the same for random component and comes with Livewire initial API request, I can see that in the network tab in dev tools.

**Reply by joshhanley** [reply]
@D-Magic ah ok. Is it always the same component that is erroring `automotive-vehicles-listing`?

**Reply by D-Magic** [reply]
@joshhanley No, on one page can randomly error different components. I missed this but it can be important that our Laravel app is working in AWS Lambda.

**Comment by joshhanley** [comment]
@D-Magic hmm ok. I don't think Lambda would be the issue. But it's hard to help you fix it without being able to consistently replicate the bug.

Typically the corruption error occurs if javascript has modified something that causes the checksum to break.

Basically if you have things like a numerically indexed array that is not sorted (like 3,1,2,...), javascript will sort them as it has no concept of associative integer arrays (as all arrays in javascript must go 0,1,2,...), but so if that happens, the checksum is expecting the data in the original order but is getting it back in the new order. So when the comparison of the hashes are done. It throws that error.

That's just one example of what could cause it. There are a couple of others, like integers that are too large and hit javascript's integer limit. I can't think of the others off the top of my head at the moment.

But so typically the way I would debug that is, if I could consistently trigger the issue, then we can log the data before the original checksum hash is generated on the initial request, and log the data from the next request that throws the checksum error, and compare the two datasets and see what has changed and then try and fix that.

Without being able to consistently replicate it though, I'm not sure how you're going to be able to track down exactly what is causing it.

I do have a tool that I wrote for helping debug this in a local development environment, but again it requires being able to consistently replicate the issue. https://github.com/joshhanley/livewire-checksum-logger

**Reply by chatisk** [reply]
What about collections then? Model collections? Do we need everytime to convert collections to array?

**Reply by joshhanley** [reply]
@chatisk standard collections and arrays are handled the same way, so you should still be able to use a standard collection and not need to convert it to an array.

Eloquent collections are a different story though. Doing things like `keyBy()` can cause issues with Eloquent collections.
